src/IssueViz/Editor/LocationTagging/LocationTagger.cs

```diff
@@ -73,25 +74,41 @@ private void OnIssuesChanged(object sender, IssuesChangedEventArgs e)
         private void UpdateTags()
         {
             var textSnapshot = buffer.CurrentSnapshot;
+            var oldTags = TagSpans;
+            TagSpans = CreateTagSpans(textSnapshot);
 
+            var affectedSpan = CalculateSpanOfAllTags(oldTags, textSnapshot);
+            NotifyTagsChanged(affectedSpan);
+        }
+
+        private List<ITagSpan<IIssueLocationTag>> CreateTagSpans(ITextSnapshot textSnapshot)
+        {
             // Get the new locations and calculate the spans if necessary
             var locations = locationService.GetLocations(GetCurrentFilePath());
-            EnsureSpansExist(locations, textSnapshot);
 
-            var oldTags = TagSpans;
-            TagSpans = locations.Select(CreateTagSpan).ToList();
+            EnsureUpdatedSpansExist(locations, textSnapshot);
 
-            var affectedSpan = CalculateSpanOfAllTags(oldTags, textSnapshot);
-            NotifyTagsChanged(affectedSpan);
+            var tagSpans = locations
+                .Where(x => x.Span.HasValue && !x.Span.Value.IsEmpty)
+                .Select(CreateTagSpan)
+                .ToList();
+
+            return tagSpans;
         }
 
-        private void EnsureSpansExist(IEnumerable<IAnalysisIssueLocationVisualization> locVizs, ITextSnapshot currentSnapshot)
+        /// <summary>
+        /// Calculate spans for locations that don't have a span, or have a span in a different text buffer. 
+        /// </summary>
+        /// <remarks>
+        /// All spans for issues in the this file should have been calculated when the file was analysed (whether primary or secondary).
+        /// However, there could be secondary locations relating to issues in other files that have not been calculated.
+        /// Also, if a file has been closed and re-opened, a new text buffer would've been created. So we need to recalculate the spans since the locations would have spans that belong to the old text buffer.
+        /// </remarks>
+        private void EnsureUpdatedSpansExist(IEnumerable<IAnalysisIssueLocationVisualization> locVizs, ITextSnapshot currentSnapshot)
         {
-            // All spans for issues in the this file should have been calculated when the file was analysed (whether primary or secondary).
-            // However, there could be secondary locations relating to issues in other files that have not been calculated.
             foreach (var locViz in locVizs)
             {
-                if (!locViz.Span.HasValue)
+                if (!locViz.Span.HasValue || locViz.Span.Value.IsEmpty || locViz.Span.Value.Snapshot.TextBuffer != currentSnapshot.TextBuffer)
```

Comment:
First repro scenario:
1. Close and reopen a file with secondary location ==> TextBuffer is different.

Second repro scenario:
1. Open a file with secondary location
2. Delete the code ==> span is empty
3. Close the file without saving & reopen ==> span is falsely empty